### PR TITLE
fix(41259) : JS autocomplete doesn't work for object literal shorthands

### LINKED ISSUE #41259: JS autocomplete doesn't work for object literal shorthands
*TS Template added by @mjbvz*

**TypeScript Version**: 4.1.0-dev.20201026

**Search Terms**
- suggest / suggestion
- intellisense
- object short hand

---

- VSCode Version: 1.51.0-insider (7a3bdf4ee9588755d447aa1c3b5db4a123fc11a9)
- OS Version: Arch Linux

Steps to Reproduce:

1. Suppose there are variables in the current context (be it local or global)
2. Create an object literal and try to use a variable in a shorthand style
3. No semantic autocompletion is offered, and textual autocomplete is only offered after 3 characters

GIF file to better demonstrate:
vscode

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by orange4glace**
Fixes #41259

Still have to work with failed testcase [`completionListAtIdentifierDefinitionLocations_destructuring`](https://github.com/microsoft/TypeScript/blob/master/tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_destructuring.ts)

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by orange4glace** [comment]
It seems that all tests should be passed to pack. I've just updated the failing one to fit the current build.

**Comment by andrewbranch** [comment]
Oh, that's annoying. I thought it just had to build cleanly. Thanks!

@typescript-bot pack this again pls

**Comment by andrewbranch** [comment]
A couple notes:

1. The failing test case is a little weird after all; the globals only show up because of the cumulative parsing weirdness that comes from the several preceding lines. My impression looking at that test was that every line was intended to be a clean parsing context. It’s valuable to ensure that a half-written statement has good completions as you’re writing it, but less so that the fourth half-written statement in a row behaves well. I would actually just split that test file up into a file per line.

2. Having an `any` or undefined contextual type is a little too restrictive. You also want to account for `object`, `unknown`, and possibly string index signatures:

    ```ts
    declare function f1(obj: unknown): void;
    declare function f2(obj: object): void;
    declare function f3(obj: Record<string, any>): void;

    declare var foo;
    f1({ /**/ });
    f2({ /**/ });
    f3({ /**/ });
    ```

    But the index signature case brings up the question of whether the types need to match up:

    ```ts
    declare const foo: string;
    const obj: { [key: string]: number } = { /* `foo` here? It would error because string is not assignable to number */ };
    ```

    On one hand, we don’t do any kind of filtering of member completions in the presence of a contextual type, but that’s more justifiable because you might dot off of one and continue the expression, resulting in a different type. More practically, it’s going to be too expensive to check the assignability of every identifier in the completions list, so I think the choice is between offering all the completions, even wrong ones, or none.

    You also want to test cases where the contextual type is a type parameter:

    ```ts
    declare function foo<T>(obj: T): T; // and another where `T extends object`, maybe `T extends {}`
    var blah: string;
    foo({ /**/ })
    ```

    I think this might Just Work once you consider `unknown` and `object` contextual types, but it’s worth testing.

Thanks for working on this!

**Comment by orange4glace** [comment]
Thanks for the much detailed feedback! It was very helpful.
1. As you've mentioned, I've modified the code so the autocompletion are also given to the `unknown`, `object`, `{}`,  `Record<string, any>` , *an object having String index type*, `<T extends object>`, `<T extends {}>`.
2. If those types are good to go, I think types with `Object` and `Object | { a: string}`(Union type including empty object type) could be also good to go. Like,
```ts
declare const foo: number;
interface Empty {}
interface Typed { typed: number; }

declare function f10<T extends Object>(obj: T): void;
declare function f13<T extends (Empty | Object | Typed)>(obj: T): void;

f10({foo});
f13({foo});
```
The last committed code contains implementation of it so I'm expecting that you can test it on the playground once you pack it.
Please give a thoughts about this suggestion!
[Playground, v4.0.5](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPCgK4C2ARiDANwBQAlihldlGAgKKkAOGAnvADeAXwZMWbBABU+3EMCHx+c4IRIUq1eKNqhIsBNmIowGeqhwBGABRJyAK0LGA1iiQB3FAEpCANyT0wHR60HA4xqbmKDgATLYOhHb24Bg+8P6BweChhhFmFtgAzPGO8ADyDilpGUG62QbhJvnR2AAsJYQASuBIMMAAPJgwjADmADTwUCh8AHzVAbUhDUZNUTgArB3w3ch9-eqUMBNTs-OZdfphK5EFAGxbgvAA2s4gfIRDowC6amSH2mdFvUrnk1tgAOwPZ6vd5EP5UH5wjQwAF+BZZS65VYFAAcWxkKkBGJyjRuLQAnP0pPAQAAPZgoYBoeBJFIzfFEi4k67NKwABipNPpIEZzIqyVM7KShCknKWIOxLUslkFdIZTKEwilCXgsrR53lWLJVhiquFovgXF4sw5+qBmNJvOwlkKZvVzOsVv48AAPuVKqZfbrZPIvNrSnr0ujaLRndZBLgkMIvHRsHEE3hk6nihmkynY+1c1nY5si-nsPcy6nIVXY3ja9hyfHE8XnXzm5ny8qO3nU5Z0y2uznB9QgA)

3. While I'm implementing this, I needed to check if the object is empty object type(ex: `{}`) or not.
Since it seems that there's no such function in `completions.ts`, I made some extra exports from `checker.ts` to use in `completions.ts` which are,
- `isTypeSubsetOf()`
- `isEmptyObjectType()`
- `globalObjectType` (`@internal getGlobalObjectType()`)

I'm not sure if it is legal to do this only for my intention.
Please give an advice whether it is okay anyway? (also, if available, looks like **typescript-bot** is complaining about the *lib update* and what can I do this for it :o)

**Comment by andrewbranch** [comment]
> While I'm implementing this, I needed to check if the object is empty object type(ex: `{}`) or not.
Since it seems that there's no such function in `completions.ts`, I made some extra exports from `checker.ts`

I wonder if you can avoid all this by seeing if `checker.getPropertiesOfType(completionsType)` returns an empty array. It would return false for non-reduced unions like `object | { a: string }`, but I actually think that’s ok—it seems reasonable to treat the property `a` in this case as an infinitely more preferable completion than random identifiers in scope. In general, I think we should start by being very cautious about flooding new completions into positions where _some_ completions are already offered. Offering new completions where _none_ are currently offered is much more clearly a win.

**Comment by orange4glace** [comment]
> it seems reasonable to treat the property a in this case as an infinitely more preferable completion than random identifiers in scope. In general, I think we should start by being very cautious about flooding new completions into positions where some completions are already offered.

Oh actually I didn't think about that! I think you are right.
I just tried `checker.getPropertiesOfType(completionsType)` and it generally works as expected including not offering global completions with union types. But also, `Object`(with capital 'O') is not getting completions while `{ [key: number]: number }` is getting them. (Actually I'm not very confident with `Object` should get completions or not?)
I'll investigate it more what I can do to filter them out :)
Thanks!

**Comment by orange4glace** [comment]
I realized that there's already a [code](https://github.com/microsoft/TypeScript/blob/master/src/services/completions.ts#L1900) which is getting properties of the type.

One thing that blocks me from implementing it is distinguishing the `Object` type since the `Object` type has some properties such as `constructor`, `toString()`, `hasOwnProperty()`.
I have compared the completion type with `globalObjectType` to achieve this which  ends up with exposing `globalObjectType` from `checker.ts`.
I think that this is the simplest way to implement it even though it introduces new *internal* API (and maybe also a low-level function as you said?)

**Comment by andrewbranch** [comment]
I don’t think we should do anything special for capital-O `Object` since it is pretty much always a mistake for people to use it. (They mean lowercase-o `object` almost 100% of the time.)

**Comment by orange4glace** [comment]
That actually makes a lot easier to me ;) Just fixed it!

File: tests/cases/fourslash/completionPropertyShorthandForObjectLiteral.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//// f12({f/*12*/});
+//// f13({f/*13*/});
+//// f14({f/*14*/});
+//// f15({f/*15*/});
+//// f16({f/*16*/});
+
+verify.completions(
+    { marker: ["1", "2", "4", "6", "10", "11", "13", "14", "15"], includes: ["foo"]},
+    { marker: ["5", "7", "8"], excludes: ["foo"], isNewIdentifierLocation: true},
+    { marker: ["3", "9", "12", "16"], excludes: ["foo"]},
```
Comment (andrewbranch) [inline tests/cases/fourslash/completionPropertyShorthandForObjectLiteral.ts:44]: If I can nitpick the format of this test a little bit, could we rename/reorder the function declarations such that they’re obviously grouped by expected behavior? So you can quickly scan that e.g. f1–f10 should show `foo` in completions, then there’s a blank line, and the next group doesn’t show `foo`, etc... We also want to add a case where `any` is the contextual type, and a case where there is no contextual type, like `const x = { /** }`.

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_destructuring.ts [hunk 2]
```diff
@@ -1,19 +1,32 @@
...
 //// var {x, y/*variable6*/
 
+// @Filename: g.ts
 //// function func1({ a/*parameter1*/
 
+// @Filename: h.ts
 //// function func2({ a, b/*parameter2*/
 
-verify.completions({ marker: test.markers(), exact: undefined });
+verify.completions({ marker: [
```
Comment (andrewbranch) [inline tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_destructuring.ts:27]: Does `test.markers()` not work cross-file or something?

**Comment by andrewbranch** [comment]
@typescript-bot pack this again please and thank you

File: tests/cases/fourslash/completionPropertyShorthandForObjectLiteral.ts (hunk 1 above)
Comment (orange4glace) [reply tests/cases/fourslash/completionPropertyShorthandForObjectLiteral.ts]: Actually I had also considered about this and I think you're right. `any` and contextual type are also added. Thanks!

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_destructuring.ts (hunk 2 above)
Comment (orange4glace) [reply tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_destructuring.ts]: Oh, I forgot to revert that. Sorry. Fixed!

**Comment by andrewbranch** [comment]
Thanks for your work on this, @orange4glace. If you’re ready for review, push the button and I’ll get some other team members to experiment with it on the playground.

**Comment by orange4glace** [comment]
I'm much appreciated that I can contribute to one of my favorite language with your lots of help :)

**Comment by andrewbranch** [comment]
We very much appreciate the contribution! Lots of people are on vacation this week, so it might not get another review until next week. This should be on track for 4.2 regardless though.

**Comment by sandersn** [comment]
I just played with this:

1. I literally thought the lack of this feature was a bug last week. I love this feature.
2. I couldn't get completions for the following examples (in TS, I didn't try JS):

Given 
```ts
const variable1 = 1
const variable2 = 2
```

```ts
// nothing after a missing comma
const o = {
  variable1
  va/**/
}

// nothing in an unclosed literal
const p = {  var/**/
interface A {
 // ...
}
```

I didn't read the original bug too closely so I'm not sure if fixing that is feature creep or not.

**Comment by orange4glace** [comment]
Thanks for the feedback @sandersn! 

> // nothing in an unclosed literal

Actually it was my mistake. I was checking `NodeFlags.ThisNodeHasError` with non-contextual type which is not necessary.
I've applied it to the latest commit.

> // nothing after a missing comma

I think it should be worked either since autocomplete works in the case which is,
```ts
const o = {
    variable1: variable1
    variable2: var/**/
}
```

I've also applied this to the latest commit.

Please leave comments for any thoughts for the code or other things!
Thanks.

File: tests/cases/fourslash/completionsSelfDeclaring2.ts [hunk 3]
```diff
@@ -9,7 +9,10 @@
 
 verify.completions({
   marker: "1",
-  exact: []
+  includes: [{
+    name: "Object",
```
[the remark this answers is not in the document]
Comment (orange4glace) [reply tests/cases/fourslash/completionsSelfDeclaring2.ts:13]: Sure! I'll fix it after I find out how I can do this elegantly. :)

> I’m almost on the fence about whether globals should show up here at all
I agree with that by the way.

**Comment by andrewbranch** [comment]
It’s probably worth discussing the broader design point of whether globals should be offered as shorthand properties, or just locals:

```ts
const salamander = 0;

const o = { s/**/ };
// Completions:
//  - salamander 👍
//  - setTimeout 🤔 
```

I think it’s a bit odd for globals to show up, but I’m not sure it’s harmful, because

- None of these show up if we have _any_ actual property suggestions from a contextual type; these only appear when we previously would have shown _no_ completions
- Locals are sorted above globals

These two points mean that the globals are not crowding out better suggestions, even if they’re rarely useful themselves. Open to hearing other takes, though.

**Comment by andrewbranch** [comment]
I just thought of something else—assuming we decide to keep globals in the completions here, can we add a test that ensures that we _don’t_ trigger auto-imports here? I think it’d be over the top to get shorthand property completions for things that aren’t even currently in scope at all.

Something like this

```ts
// @module: esnext

// @Filename: /a.ts
//// export const exportedConstant = 0;

// @Filename: /b.ts
//// const obj = { exp/**/

verify.completions({
  marker: "",
  exact: completion.globals, // I think?
  preferences: { includeCompletionsForModuleExports: true }
});
```

**Comment by orange4glace** [comment]
> I think it’d be over the top to get shorthand property completions for things that aren’t even currently in scope at all.

I agree that offering auto-imports to completions might be annoying 👍 
While I have implemented it, I realized that autocompletion includes its container variable, for example,
```ts
const obj = {
    o/**/ 
     ~~~~ obj
}
```
It is obviously a mistake to use its container variable (and which finally emits an error `Block-scoped variable 'obj' used before its declaration.`) so technically, I think we should get rid of it from completions.
On the hand, *property value assignment* also auto-completes its container variable like,
```ts
const obj = {
    value: o/**/
            ~~~~ obj
}
```
[Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBCBGArGBeGBvAUDHMBuAhgDYCuApgFxyYC+QA)

I don't know whether this is intended or not but if we have to get rid of this, implementing it seems quite cumbersome I think since there's such a case like,
```ts
const outer = {
    inner: {
        nested: {
            value: o/**/
        }
    }
}
```
then we have to check all the parent node until we find the `outer` node which might lead to some performance issue. (There might be a better solution for this but I'm not a pro of Typescript codebase so.. 🤔)



> It’s probably worth discussing the broader design point of whether globals should be offered as shorthand properties, or just locals:

👍  for not offering global symbols since as you said, literally no one wants to re-declare such symbols which can be accessed globally.

**Comment by andrewbranch** [comment]
> I don't know whether this is intended or not (like implementing it is kinda cumbersome? 🤔)

I don’t think it should be difficult to fix; I’m guessing it has just gone unnoticed. But I don’t think you need to worry about it in this PR since it appears unrelated. I can open a new bug for it.

File: tests/cases/fourslash/completionsSelfDeclaring2.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (orange4glace) [reply tests/cases/fourslash/completionsSelfDeclaring2.ts:13]: Fixed!

[the remark this answers is not in the document]
Comment (orange4glace) [reply tests/cases/fourslash/completionsSelfDeclaring2.ts]: Fixed! except it uses `globalsPlus` for this case (https://github.com/microsoft/TypeScript/issues/41731)

**Comment by andrewbranch** [comment]
@sandersn @DanielRosenwasser either of you want to weigh in on whether these completions should include globals  (https://github.com/microsoft/TypeScript/pull/41539#issuecomment-733864143)? I don’t feel super strongly about it for the reasons I outlined in that comment. But I want to make a decision and get this merged soon.

**Review by andrewbranch — APPROVED** [review]
Thanks @orange4glace!

**Comment by orange4glace** [comment]
Oh, I haven't noticed that this merged.
Thanks @andrewbranch! I couldn't get this done if there wasn't your help. ;)

**Pull request merged**